### DISCUSSION TITLE
Como vocês trabalham com WordPress?

### DISCUSSION THREAD

**Discussion by kvnol** [frontendbr/forum#577] [Discussão]
Muitas dúvidas foram sanadas com as issues #17 (Bower ou NPM) e #54 (Como é seu workflow?), mas gostaria de levantar uma discussão sobre WordPress. Como não utilizo muito o WordPress em projetos, quando aparece alguma coisa eu faço como um bom homem das cavernas, arcaicamente.

Sabendo disto, gostaria de saber como é o workflow de vocês com o WordPress: o que utilizam, como/o que versionam e como fazem deploy disso tudo (só o tema ou todo o WordPress?).

**Comment by Cleysonlb** [comment]
Ultimamente não tenho mexido muito com Wordpress, mas quando tenho que desenvolver algo uso Gulp, stylus, Pug (jade), e lints normal do workflow, mas configuro para a saida dos arquivos irem direto para um um "blank template" no wordpress onde começo a desenvolver em cima.

**Comment by kvnol** [comment]
@Cleysonlb tu usa pug pra gerenciar as views, mas e o php do WordPress?

**Comment by raulfernando** [comment]
@kvnol essa tbm era minha duvida quanto ao PUG 

Aqui no trabalho usamos bastante o wordpress e oq temos feito é usar o docker-compose assim desenvolvemos usando gulp e a pasta de saida colocamos como um volume no docker direto para a pasta themes

**Comment by theandersonn** [comment]
@kvnol Tbm tenho pesquisado e tentando melhorar o meu fluxo de desenvolvimento.

Ainda não cheguei no resultado que desejo:

- Versiono todo o projeto incluindo o core, tema e plugins; isso tudo vai para um repositório privado; 
- Dentro de wp-content incluo backup do banco de dados;
- No tema eu separo pedaços reaproveitáveis na pasta 'templates-parts';
- Divido as funções que normalmente vão em functions na pasta 'inc';
- Carrego estilos e scripts com a função enqueue_scripts();
- O menu eu monto com funções e faço chamada no front-end;
- O JS normalmente utilizo jquery, concateno com libs externas;
- Uso SASS com smacss;
- Em algumas situações utilizo o grid do bootstratp e alguns componentes e em outras não;
- Para gerenciar as tarefas, utilizo gulp;
- Deploy para VPS [ex: digital ocean] faço através do github (preciso melhorar esse aspecto, ver heroku, testes, explorar outras possibilidades);
- Em alguns projetos ainda utilizo ftp;

Referências disso tudo que falei.

- [Tema](https://github.com/theandersonn/blank-theme-wp);
- [Gulp](https://github.com/theandersonn/gulp-dynamic);
- [Bootstrap Smacss](https://github.com/theandersonn/bootstrap-smacss);
- [Snippets do dia dia](https://github.com/theandersonn/wp-snippets).

**Comment by nandomoreirame** [comment]
@kvnol eu costumo desenvolver o tema do zero mesmo, criando o front-end (inicial) primeiro e depois criando a estrutura WordPress. Eu uso o core do [Odin](https://github.com/wpbrasil/odin) para algumas coisas, classes e funções úteis.

Hoje trabalho com Docker e Docker-compose para automatizar a instalação do WordPress na minha máquina, não necessitando instalar o PHP, MySQL, Apache e phpmyadmin localmente.

[Montei uma estrutura base para dar start nos projetos usando docker](https://github.com/nandomoreirame/docker-wordpress), eu instalo o que preciso nos containers do docker para rodar o WordPress e o phpmyadmin.

Quanto a estrutura de front estou usando, sass como pre-processador e gulp como task runner :smiley:

**Comment by gabrieldarezzo** [comment]
@claudiosanches alguma dica?

**Comment by claudiosanches** [comment]
@gabrieldarezzo muitos anos que não faço mais qualquer freela para cliente entregando tema ou site pronto assim, mas quando fazia não gostava de versionar o projeto inteiro, geralmente versionava os plugins e tema que desenvolvia em repositórios separados, até mesmo para poder reaproveitar meu trabalho depois. Além que tinha um repositório com submodules do meus repositórios, apenas para versionar o estado atual do projeto e o wp-config.php, que era o que usava para deploy, o resto já teria que permanecer sempre atualizado com os updates do WordPress.
Hoje também rola usar wp-cli no servidor, quando o santo usa um VPS pelo menos, assim já da para instalar, manter atualizado e até mesmo verificar por arquivos hackeados/corrompidos no core.

Mas tem um puta tempo que não faço isso, até mesmo freela deve ter uns 4 anos que não faço nenhum e nos últimos anos que fiz freela foram tudo apenas com plugins, fazia um ou mais plugins para o cliente, ele pegava e instalava, e quando não funcionava, o cliente me dizia e eu corrigia, assim deixava funcionando sem receber acesso de absolutamente nada, o que era uma maravilha, porque não tem nada mais chato do que ter acesso de servidor de terceiros.

**Comment by felipefialho** [comment]
@Mendrone rei do WordPress

**Comment by giioohbernini** [comment]
@kvnol Muito do que eu faço já foi falado ai pelo @raulfernando e o @theandersonn, mas um tempo atrás também me esbarrei nesse assunto, e consegui tirar muita coisa boa desse repo [aqui](https://github.com/optimized/docker-wordpress-gulp-starter-kit), pode não estar tão mastigado, mas na época me ajudou bastante

**Comment by ghost** [comment]
Atualmente eu só trabalho com WordPress, o tema eu sempre utilizei o do próprio _Automattic_ [Underscores](http://underscores.me). Excluo algumas coisas irrelevantes e uso o meu próprio workflow, [Stag Starter](https://github.com/wdss93/estag-starter).

Recomendo dar uma olhada nesses temas: [untheme](https://github.com/taniarascia/untheme) e [startwordpress](https://github.com/taniarascia/startwordpress).

Não são repositórios famosos, mas a criadora [Tania Rascia](https://github.com/taniarascia) é alguém que vive WordPress.

**Comment by marioernestoms** [comment]
- no ambiente local uso o https://github.com/Varying-Vagrant-Vagrants/VVV
- faço a instalação do WordPress por submodules.
- versiono somente os plugins e o tema
- tema geralmente uso o http://underscores.me/ 
- uso o Gulp para gerenciar e o Bower para incluir as bibliotecas externas de CSS e JS
- no CSS uso Sass + Compass 
- Para criar custom fields, metaboxes e etc uso ou o ACF ou o Carbon Fields e coloco em um plugin. 
- Para deploy to usando o Jenkis pra fazer deploy automatico pelo github (onde geralmente tenho 3 branchs Master (Produção), Stagging (Homologação) e develop (Desenvolvimento).

**Comment by rubenmarcus** [comment]
legal e o que o srs. acham de frameworks SPA com temas do wordpress?

**Comment by omariosouto** [comment]
Cara eu uso um tema que eu criei, nele eu tenho ES6 com Standard como linter, Stylus no CSS e só kk.
- Criei o tema baseado no underscore;
- Uso uma arquitetura de componentes estilo react pra fazer include dinâmico das dependências do projeto;
- Faço critical rendering path;
- Dou suporte pra http2 e http1 fica de acordo com a verba do cliente;
- E algumas mágicas;

A documentação não está pronta (vou fazer), mas estou extremamente satisfeito com o tema, no nível de eu copiar o menu de um projeto, colar no outro e conseguir realmente reaproveitar código.

Segue link ai, quem quiser entender como funciona só me chamar ;)
https://github.com/soutomario/wp-fenrir

**Comment by renanlara** [comment]
Cara eu uso o [Odin](https://github.com/wpbrasil/odin), nele eu editei algumas coisas do Front-End e acrecei poucas coisas do Back-End. O próprio Odin vem com bastante coisa, como Bootstrap, SASS, GruntJS e outras coisas. 

Preciso só arrumar uma solução para deploy, de resto estou muito satisfeito.

**Comment by joaodfmota** [comment]
WordPress project boilerplate 
[Bedrock](https://github.com/roots/bedrock) 
WordPress starter theme
[Sage](https://github.com/roots/sage)

Recomendo ler a documentação ou acessar o site [roots.io](roots.io), eles explicam melhor o uso e utilidades das ferramentas que citei acima.

No sage eu faço algumas modificações( gulp, e  setup do tema conforme a necessidade da empresa/projeto)

**Versionamento:**
Como 99,9% dos projetos que trabalho são conversão de AI/PSD para WP então eu só versiono a pasta do tema.

**Comment by phsantiago** [comment]
@theandersonn 
"Dentro de wp-content incluo backup do banco de dados;"
Cuidado com isso cara, tenha certeza que seu servidor de aplicação não está servindo arquivos com a extensão do seu backup, pois a wp-content é acessível e backups são considerados arquivos estáticos.
Se alguém acertar o filename pode baixar seu backup.

**Comment by theandersonn** [comment]
@phsantiago Obrigado pelo aviso. 
Esse backup fica somente no ambiente de desenvolvimento, não vai para produção. Como citei na época ainda não é o ideal, desejo automatizar algumas coisas.

**Comment by vitorebatista** [comment]
Como que vocês trabalham com _i18n_ de posts? Existe algum plugin ou vocês duplicam o wordpress inteiro para outros idiomas?

**Comment by theandersonn** [comment]
Recomendo esses passos:
- [Como inserir suporte a tradução em temas e plugins do WordPress](https://www.rafaelfunchal.com.br/blog/tutoriais/como-inserir-suporte-a-traducao-em-temas-e-plugins-do-wordpress/)
- [Como traduzir temas e plugins do WordPress](https://www.rafaelfunchal.com.br/blog/tutoriais/como-traduzir-temas-e-plugins-do-wordpress/)

**Comment by vitorebatista** [comment]
@theandersonn  obrigado pelo retorno!

A princípio eu preciso de um plugin para gerenciar traduções exclusivamente para posts (é um blog de empresa). Vi que ![WMPL](https://wpml.org/pt-br/) e o ![QTranslate-X](https://br.wordpress.org/plugins/qtranslate-x/) fazem isso, vou iniciar uns testes mas queria ver se alguém já passou por esta necessidade.

Abs.

**Comment by theandersonn** [comment]
Beleza! Na questão textual já utilizei o [polylang](https://br.wordpress.org/plugins/polylang) e atendeu de boa.

**Comment by phsantiago** [comment]
A parte legal do polylang é que ele tem uma integração ótima com o [ACF](https://www.advancedcustomfields.com/)

**Comment by lazincastro** [comment]
Para o pessoal que usa o docker-compose com wordpress, como que vocês fazem a migração do banco de dados?

**Comment by raulfernando** [comment]
Criei esses scripts para fazer esse trabalho de export/import

[Export](https://gist.github.com/raulfernando/3953b8417f3be3be632bd198c40cb4e7)/[Import](https://gist.github.com/raulfernando/6a0b6ab4fbcf194781894dec0f2e6292)


--

**Comment by lazincastro** [comment]
Esqueci de colocar na minha pergunta anterior, mas eu tenho outra dúvida, estou usando a configuração recomendada pelo próprio site do docker https://docs.docker.com/compose/wordpress/ funciona normalmente.

Depois que termino de fazer algumas alterações em um tema, eu paro com "docker-compose down --volumes" dai quando inicio "docker-compose up -d" ele volta pra tela de instalação do wordpress. 

Eu gostaria que os dados (postagens, paginas, etc) permanecessem para que depois de ter feito tudo eu exporte para o servidor de produção por exemplo.

Alguém saberia me ajudar?

**Comment by JoseCage** [comment]
@lazarocastro como está seu `docker-compose.yml`? Tem que configurar a persistência de dados no seu docker-compose.

Leia um pouco sobre [volumes](https://docs.docker.com/engine/admin/volumes/volumes/)..

**Comment by raulfernando** [comment]
Pra isso vc precisa criar um volume no contêiner do MySQL com os dados do banco. Eu faço conforme  exemplo abaixo. 

    volumes:
      - ./.data/db:/var/lib/mysql
